### DISCUSSION TITLE
Help creating new workspace templates with Terraform plans

### DISCUSSION THREAD

**Discussion by kiebrew** [microsoft/AzureTRE#3870] [Q&A] [answered]
Hi there, we're looking to learn the process for authoring new Workspace Templates based on the guidance [here](https://microsoft.github.io/AzureTRE/v0.16.0/tre-workspace-authors/authoring-workspace-templates/). We wanted to do a super simple workspace template: Base + Azure SQL Server & Database.

I've written my Terraform code, but I want to be able to run successful Terraform plans before bundling everything up in porter, registering with the API etc. 

How are people doing local Terraform development for creating new templates? I am struggling to pass the correct parameters when doing TF plans. There is a .env sample file which contains some values that makes sense and some that don't. Is it just a case of building a .tfvars file with the required Input variables? I feel that I am missing a trick somewhere, any help or guidance would be greatly appreciated.

For some additional context, we have cloned the AzureTRE-Deployment repo and are running from the local devcontainer in VSCode.

**Comment by marrobi** [comment, marked as answer]
Hi @kiebrew , thanks for getting in touch. We could do with some better documentation in this space as per #3258 , hopefully something we can progress in the coming weeks.

My workflow is to put a `deploy.sh` file into the terraform directory (see an example `templates/workspace_services/guacamole/terraform/deploy.sh`), and then run `make terraform-deploy`, this imports the contents of the `.env` file into the appropriate `TF_VAR environment variables`. The reason we use `.env` is it can also be used with Porter.

Please be aware any terraform operations  that needs data plane connectivity is likely to fail as the resources are on a VNET. I have dabbled with connecting VSCode to the resource provider using https://code.visualstudio.com/docs/remote/tunnels , but haven't got round to documenting it. This gives the closest experience to an actual deployment.

Let me know if that helps at all.

**Reply by kiebrew** [reply]
Thanks for taking the time to reply @marrobi . This does indeed help, I'll bear those in mind as I get deeper into learning the TRE deployment mechanisms. Thank you 😄
